### ISSUE TITLE
RESTfull Contact API

### ISSUE THREAD

**Issue by themouette** [CaliOpen/caliopen.api#7]
Contact API is not RESTfull yet.

I propose we use the following interface:

## GET /api/v1/contacts

Returns a collection of contacts.

The collection is wrapped in a `contacts` namespace. Metadata should be at the response root as well. Sample response:

``` json
{
  "total": 300,
  "contacts": [
    {"contact_id": "xyz", "firstname": "John", "lastname": "Doe"}
  ]
}  
```

## GET /api/v1/contacts/:contact_id

Return a Contact resource identified by `:contact_id`

The contact is wrapped in a `contacts` namespace. Metadata should be at the response root as well. Sample response:

``` json
{
  "contacts": {"contact_id": "xyz", "firstname": "John", "lastname": "Doe"}
}
```

## POST /api/v1/contacts

Create a new contact from request payload.

Returns **201** status when everything went fine

The updated contact is wrapped in a `contacts` namespace. Metadata should be at the response root as well. Sample response:

``` json
{
  "contacts": {"contact_id": "xyz", "firstname": "John", "lastname": "Doe"}
}
```

**Note** that response MUST return a link to newly created resource in the `Location` header. Sample Response:

```
HTTP/1.1 201 Created
Content-Length: length
Cache-Control: no-cache
Connection: keep-alive
Content-Type: application/hal+json
Date: Fri, 30 Jan 2015 13:08:46 GMT
Location: http://caliopen.org/api/v1/contacts/xyz
```

## PUT /api/v1/contacts/:contact_id

Update a contact  resource identified by `:contact_id`

Returns **200** status, when everything went fine.

The updated contact is wrapped in a `contacts` namespace. Metadata should be at the response root as well. Sample response:

``` json
{
  "contacts": {"contact_id": "xyz", "firstname": "John", "lastname": "Doe"}
}
```

> More about PUT response status http://stackoverflow.com/a/827045

## DELETE /api/v1/contacts/:contact_id

Delete Contact resource identified by `:contact_id`

Returns **204** status,  with an empty body.

## PATCH /api/v1/contacts/:contact_id

Not Implemented Yet

## Error Handling

Errors relies on HTTP codes.

* `400`: Validation error
* `401`: User is not authenticated, resource is protected
* `403`: User is authenticated, resource is protected
* `404`: Resource not found

Error resource is wrapped in an 'error ' namespace.

### Sample application error:

``` json
{
  "error": {
    "code": "500",
    "name": "Internal Error",
    "message": "An internal error occured"
  }
}
```

### Sample validation error:

``` json
{
  "error": {
    "code": "400",
    "name": "Bad Request",
    "message": "Provided data is not valid",
    "globals": [ "You cannot create resources" ],
    "fields": {
      "firstname": ["This field is required"]
    }
}
```

**Comment by gdchamal** [comment]
Agree on most point, but not sure about namespace usage when getting a single object. It's not in habits I'm aware of.

**Comment by themouette** [comment]
Indeed, namespacing on single resource is not mandatory, but makes ember integration easier from what I see.

It can be removed though, I'll find a workaround. The only requirement is to be consistent on all of our APIs.

**Comment by themouette** [comment]
What do you think about keeping [VCard properties](http://en.wikipedia.org/wiki/VCard#Properties) as a reference for possible fields?

It makes API learning curve easier and ease integration with external tools.

**Comment by gdchamal** [comment]
vcard is only a serialization format (using content-type). Contact structure is mostly based on this specification : https://developers.google.com/gdata/docs/2.0/elements?csw=1#gdContactKind and contains all data needed for a vcard import / export
